pkg/storegateway/batch_series.go

```diff
@@ -77,110 +55,69 @@ func openBlockSeriesChunkRefsSetsIterator(
 		stats.merge(&unsafeStats)
 	}
 
-	return &blockSeriesChunkRefsSetsIterator{
-		blockID:                    blockID,
+	postingsIterator := newPostingsSetsIterator(ps, batchSize)
+
+	inflatingIterator := &inflatedSeriesChunkRefsSetIterator{
+		ctx:                 ctx,
+		postingsSetIterator: postingsIterator,
+		indexr:              indexr,
+		stats:               stats,
+		blockID:             blockID,
+		shard:               shard,
+		seriesHasher:        cachedSeriesHasher{cache: seriesHashCache},
+		skipChunks:          skipChunks,
+		minTime:             minTime,
+		maxTime:             maxTime,
+	}
+
+	limitingIterator := &limitingSeriesChunkRefsSetIterator{
+		from:          inflatingIterator,
+		chunksLimiter: chunksLimiter,
+		seriesLimiter: seriesLimiter,
+	}
+	return limitingIterator, nil
+}
+
+type postingsSetsIterator struct {
+	postings []storage.SeriesRef
+
+	batchSize                  int
+	currentBatchPostingsOffset int
+	currentSet                 []storage.SeriesRef
+}
+
+// TODO dimitarvdimitrov add filterPostingsByCachedShardHash to this constructor
+func newPostingsSetsIterator(postings []storage.SeriesRef, batchSize int) *postingsSetsIterator {
```

Comment:
I think it's fine calling it "batch" here because that's exactly what we're doing. Given a list of all postings, we batch it.

```suggestion
func newPostingsBatchedIterator(postings []storage.SeriesRef, batchSize int) *postingsSetsIterator {
```

Reply:
my reasoning was that it iterates and produces sets of postings. It's similar naming that we've used for `seriesChunksSetIterator` and `seriesChunkRefsSetIterator`, except this one isn't an `interface`